sd: omit zero-initialization for local variable and add config SD_CMD_RETRIES

ndrs-pst opened the pull request:
This series of commits is an effort to remove zero-initialization, specifically targeting the `sdhc_command` and `sdhc_data` structures.
Suggestions and insights from maintainers are always welcome and appreciated. 🚀

ndrs-pst committed:
sd: omit zero-initialization for sdhc_command and sdhc_data structures

The decision to omit zero-initialization is driven by a desire to enhance
Zephyr's compactness and efficiency.

This is achieved by omitting zero-initialization, thereby reducing the
instruction count and, as a byproduct, the code size.

After a thorough review of the usage of struct sdhc_command and sdhc_data,
it has been determined that zero-initialization can be omitted.
Only a portion of the fields need to be manually initialized.
(e.g. cmd.retries, data.block_addr)

For the uninitialized fields, it can be expected from successful
operations that data will be appropriately written back from
the underlying layer.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

ndrs-pst committed:
sd: omit zero-initialization for csd[4] and cid[4]

After a review of the usage of csd[4] and cid[4],
it has been determined that zero-initialization can be omitted.

It can be expected from successful operations that data will be
appropriately written back from the underlying layer.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

ndrs-pst committed:
sd: add config SD_CMD_RETRIES

Introduced the config SD_CMD_RETRIES option as a build-time parameter.
This allows the assignment of a specific value to cmd.retries.
The default value has been set to 0 to preserve the previous setting.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

decsny commented:
can I ask why?

ndrs-pst commented:
@decsny, From my point of view, the decision to omit zero-initialization from the `sdhc_command` and `sdhc_data` structures is driven by a desire to enhance Zephyr's compactness and efficiency.

I acknowledge that the definitions of 'compact' and 'efficient' can vary from person to person.
However, in the context of Zephyr, I believe this change contributes to reducing unnecessary operations, thereby streamlining the codebase.

decsny commented:
> @decsny, From my point of view, the decision to omit zero-initialization from the `sdhc_command` and `sdhc_data` structures is driven by a desire to enhance Zephyr's compactness and efficiency.
> 
> I acknowledge that the definitions of 'compact' and 'efficient' can vary from person to person.
> However, in the context of Zephyr, I believe this change contributes to reducing unnecessary operations, thereby streamlining the codebase.

Can you just put in the commit message the reason you're doing it, like if you think it reduces instruction count or any other reason

ndrs-pst commented:
> Can you just put in the commit message the reason you're doing it, like if you think it reduces instruction count or any other reason

@decsny, I have revised the commit message.
Could you please review the updated message at at https://github.com/zephyrproject-rtos/zephyr/pull/68017/commits/f7406d65fd33b428ab18b8a40d05e5728a3db1e0 ? Thank you.

danieldegrasse commented on subsys/sd/sd_ops.h:
```diff
@@ -10,7 +10,7 @@
 /*
  * Switches voltage of SD card to 1.8V, as described by
- * "Signal volatage switch procedure" in section 3.6.1 of SD specification.
+ * "Signal voltage switch procedure" in section 3.6.1 of SD specification.
```
Thanks for catching this- while you are at it, it looks like the section is now named "Signal Voltage Switch Sequence". We should also drop the section number, that may change with revisions of the specification

danieldegrasse replied:
I was indeed looking at the wrong spec...apologies, shouldn't be commenting on GitHub so early in the morning 😆. Could you change "SD specification" to "SD host controller specification" to protect me from myself? I do want to refer the section 3.6.1 in the SD host controller simplified specification

ndrs-pst replied:
Already push another commit

danieldegrasse reviewed (commented):
This does save a decent amount of flash space- when building `tests/subsys/sd/sdmmc` on the RT1060, utilized flash space goes from 71168 Bytes on sha 04de43b1a119665 to 71016 Bytes on sha c716b805 (head of this PR)

ndrs-pst committed:
sd: update comment to reference 'SD host controller specification'

In the code comments referring to section 3.6.1, changed the reference
from 'SD specification' to 'SD host controller specification'
for a clearer and more accurate context.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>
